versioned/storage/store/src/main/java/org/projectnessie/versioned/storage/versionstore/CommitImpl.java

```diff
@@ -186,11 +199,31 @@ CommitResult<Commit> commit(
 
     fromCommitMeta(metadata, commit);
 
+    CommitObj newHead;
     try {
-      CommitObj newHead = commitLogic.doCommit(commit.build(), objectsToStore);
-
+      CreateCommit createCommit = commit.build();
+      newHead =
+          commitLogic.buildCommitObj(
+              createCommit,
+              c -> CONFLICT,
+              (k, v) -> {},
+              NO_VALUE_REPLACEMENT,
+              NO_VALUE_REPLACEMENT);
+
+      // If 'commitRetryState.storedContents' already contains the commit-ID, __we__ already
+      // successfully persisted that commit. This can happen, if the `Persist` implementation raised
+      // an `UnknownOperationResultException` in one of the following operations (reference bump for
+      // example).
+      CommitObj stored = commitLogic.storeCommit(newHead, objectsToStore);
+      if (stored != null) {
+        newHead = stored;
+        commitRetryState.commitPersisted = newHead.id();
+        commitRetryState.storedContents.addAll(toStore);
+      }
+      boolean commitPersisted =
+          stored != null || newHead.id().equals(commitRetryState.commitPersisted);
```

Comment:
> why even attempt a duplicate storeCommit call?

We have to do something here, if we get here again - either the commit, related objects or the reference weren't properly persisted. But we do not exactly now what and why and whether sth was persisted in case of an `UnknownOperationResultException`.

I suspect these timeouts to be rare - so would like to not add to much code to handle these cases (but complicate the usual good cases).

Reply:
good point :+1: